### PR TITLE
compilation fixes

### PR DESCRIPTION
**Pull request by hannesm**
best viewed commit by commit. this is supposed to achieve the following (by cherry-picking and polishing commits from a multitude of forks):
- use dune instead of ocamlbuild/topkg
- revert the "remove sexplib serialisers"
- use `dune-configurator` and `bigarray-compat`
- do not use `-Wno-implicit-fallthrough`, since this is a reasonable warning. instead, mark the case statements explicitly where a fallthrough is expected (use a C comment instead of annotation to make gcc<7 and clang not complain)
- fix "unused C function" warnings
- cross-compilation for freestanding and xen (by flattening the `src/native` directory and copying around)
- be more precise why `warning -3` is in nocrypto_lwt_entropy

the freestanding and xen support ~~is not yet properly tested (and may need some more work)~~ works fine (a static_website_tls unikernel links, and the freestanding/hvt one also runs smoothly), I'll follow up on that. once this is good to go, the renaming could happen.

### CODE REVIEW

File: dune-project [hunk 1]
```diff
@@ -0,0 +1,3 @@
+(lang dune 1.7)
+(name nocrypto)
+(version %%VERSION_NUM%%)
```
Comment (emillon) [inline dune-project]: this is not necessary, `dune subst` will populate that when pinning and `dune-release` will insert it at release time.

File: config/cfg.ml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+let flags = ["-DACCELERATE"; "-mssse3"; "-maes"; "-mpclmul"]
+
+let _ =
+  let auto = match Cpuid.supports needs with Ok true -> flags | _ -> [] in
+  let fs = match Sys.getenv evar with
+    "true" -> flags
+  | "false" -> []
+  | _ -> auto
+  | exception Not_found -> auto in
+  Format.(printf "(@[%a@])%!" (fun ppf -> List.iter (fprintf ppf "%s@ ")) fs)
```
Comment (emillon) [inline config/cfg.ml]: You can use Configurator.V1.Flags.write_sexp here.

Comment (hannesm) [reply config/cfg.ml]: ok, I do not know how to do this. I am unable to find the documentation / API of `Configurator.V1.Flags`. this `cfg.exe` is used by `src/dune` (with a rule `(rule (with-stdout-to c_flags.sexp (run ../config/cfg.exe)))`) as well as xen/dune and freestanding/dune.

From https://dune.readthedocs.io/en/stable/quick-start.html#defining-a-library-with-c-stubs I gather that your proposed `write_sexp` takes two arguments, a file name and an OCaml value - but it is rather unclear to me where the output is written to, then (in terms of relative path).

File: .gitignore [hunk 3]
```diff
@@ -2,11 +2,15 @@ _build
 *.install
```
Comment (avsm) [inline .gitignore]: can remove the  (native, byte) from here as dune places those in _build

File: .ocamlinit [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-#directory "_build/src"
-#load "nocrypto.cma"
-
-(* #require "oUnit" *)
-(* #directory "_build/tests" *)
-(* #load "testlib.cma" *)
+#directory "_build/default/src"
+#directory "_build/default/src/.nocrypto.objs/byte"
+#directory "_build/default/unix"
+#directory "_build/default/unix/.nocrypto_entropy_unix.objs/byte"
```
Comment (avsm) [inline .ocamlinit]: This file shouldn't be needed now either, as `dune utop` should do the trick to generate the toplevel

File: .travis.yml [hunk 5]
```diff
@@ -7,7 +7,6 @@ env:
     - UPDATE_GCC_BINUTILS=1
     - TESTS=true
   matrix:
-    - OCAML_VERSION=4.02
     - OCAML_VERSION=4.03
     - OCAML_VERSION=4.04
     - OCAML_VERSION=4.05
```
Comment (avsm) [inline .travis.yml]: We can remove everything <4.06 from here I believe...

File: lwt/nocrypto_entropy_lwt.ml [hunk 6]
```diff
@@ -1,19 +1,23 @@
-open Lwt
+open Lwt.Infix
 open Nocrypto
 
+[@@@ocaml.warning "-3"]
+(* Used for hooks, https://github.com/ocsigen/lwt/issues/573 *)
+module Lwt_sequence = Lwt_sequence
```
Comment (avsm) [inline lwt/nocrypto_entropy_lwt.ml]: This can be Lwt_dllist as well, but that can happen in a subsequent PR

File: nocrypto.opam [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+license:      "ISC"
+synopsis:     "Simple crypto for the modern age"
+
+build: [ ["dune" "subst"] {pinned}
+         ["dune" "build" "-p" name "-j" jobs ]
+         ["dune" "runtest"] {with-test} ]
+
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {build & >= "1.7"}
```
Comment (avsm) [inline nocrypto.opam]:
```suggestion
  "dune" {>= "1.7"}
```

File: .gitignore (hunk 3 above)
Comment (hannesm) [reply .gitignore]: thanks, done

File: dune-project (hunk 1 above)
Comment (hannesm) [reply dune-project]: thanks, done

File: .ocamlinit (hunk 4 above)
Comment (hannesm) [reply .ocamlinit]: ok, done

File: .travis.yml (hunk 5 above)
Comment (hannesm) [reply .travis.yml]: I actually removed travis.yml, and we'll focus on ocaml-ci here :)

File: nocrypto.opam (hunk 7 above)
Comment (hannesm) [reply nocrypto.opam]: already done :)

**Comment by hannesm** [comment]
thanks for your reviews.

I force-pushed to add a lower bound of OCaml 4.07.0 -- and then removed the `bigarray-compat` commit (since this is only needed < 4.07.0 -- in 4.07.0+ we have Stdlib.Bigarray)

File: lwt/nocrypto_entropy_lwt.ml (hunk 6 above)
Comment (hannesm) [reply lwt/nocrypto_entropy_lwt.ml]: thanks; please take a look into 45e7a86 for an alternative solution -- no deprecated API, no dllist

**Comment by hannesm** [comment]
hmm, now when I run `dune utop`, I get:
```
File "xen/dune", line 7, characters 12-28:
7 |  (libraries mirage-xen-posix)
                ^^^^^^^^^^^^^^^^
Error: Library "mirage-xen-posix" not found.
Hint: try: dune external-lib-deps --missing .utop/utop.exe
```

but in `xen/dune` the stanza `optional` is included -- any dune hints how to get `dune utop` to work in such a scenario (dune 2.1.3 is what I use)?

another question about CI: is it possible that it runs with optional dependencies? (in case the answer is no (what I expect), I guess an appropriate solution is to have separate packages (though I'm not sure how this would work with the `META.in`, but I can try to figure that out).

**Comment by talex5** [comment]
> another question about CI: is it possible that it runs with optional dependencies? (in case the answer is no (what I expect), I guess an appropriate solution is to have separate packages (though I'm not sure how this would work with the META.in, but I can try to figure that out).

Yeah, separate packages would be great!

If you need to force optional deps to be installed in CI as a quick fix though you could always add a dummy opam package that depends on all the optional stuff. Or you could make the optional dependencies required for tests.

**Comment by hannesm** [comment]
@talex5 thanks for your quick and valuable response. Since I'm not eager to accumulate hacks in this repository, but will subsequently revise the package layout (not in this PR).

**Pull request merged**